### PR TITLE
211#6 Update code to use new products interface.

### PR DESCRIPTION
**Pull request by jsolbrig**
This PR is now ready for review!

1. DONE Update data_fusion procflow
2. DONE test_interfaces fixed (for geoips tests)
3. DONE output_dict updates made (for geoips tests with product_params_override)
4. DONE Finalize release notes

https://github.com/NRLMMD-GEOIPS/geoips/pull/221/files#diff-8389e2591ef69b0939a0af0b2ee3f9a187737a5311a88b54f4ba31502e45bdb5

# Reviewer Checklist
https://github.com/NRLMMD-GEOIPS/.github/blob/main/.github/review-template.md

# Related Issues
fixes NRLMMD-GEOIPS/geoips#211

# Testing Instructions
$GEOIPS/tests/test_full_install.sh

This must be fully functional!

# Summary

Please see release notes for full discussion of changes:

https://github.com/NRLMMD-GEOIPS/geoips/pull/221/files#diff-8389e2591ef69b0939a0af0b2ee3f9a187737a5311a88b54f4ba31502e45bdb5

Version 1.10.0a12 (2023-05-09)
******************************

Overall YAML interface updates

* Remove old YAML discovery functions in geoips_utils
* Remove old style YAML configs
* Update dev/output_config.py to use new YAML interfaces
* Rework mapping routines to use new interfaces

Feature and gridline annotators interface

* Add feature and gridline annotators YAML-based interface
* Refactor gridline_annotator and feature_annotator plugin schema and plugins
* Update boundaries/gridlines_info to feature/gridline_annotator
* Update boundaries/gridlines_params to feature/gridline_annotator

Sectors interface

* Add new sector YAML-based interface
* Remove sectorfiles option from commandline

Dynamic sectors interface

* Update sector YAML-based interface to suppose dynamic sectors.
* Update all TC templates to use updated sectors.generated schema
* Update command line argument from tc_template_yaml to tc_spec_template

Sector adjusters interface

* Add sector_adjusters module-based interface
* Update single_source procflow to use new sector_adjusters interface
* Replace adjust_area_def command line arg with sector_adjuster

Colormaps interface

* Add colormap module-based interface
* Add plugins/\*\*.txt to pyproject.toml
* Move colormaps from geoips/image_utils/ascii_palettes to
  plugins/txt/ascii_palettes
* Update matplotlib_linear_norm to be a "configurable" colormap

  * Add "find_ascii_palette" routine to geoips_utils.py
  * Allow cbar_ticks == None to plot min/max of actual data
  * Allow matplotlib, ascii, and geoips colormap sources
* Change colormap families to "matplotlib"
* Run pytest on all installed geoips plugin packages

Products interface

* Add products YAML-based interface
* Allow passing dictionary of product_spec_overrides to products.get_plugin
* Support multiple coverage checkers for a single product
* Testing updates

  * Add data_fusion stitched test to test_full_install.sh
  * Add template_basic_plugin test to test_full_install.sh
* Bug fixes

  * Fix plugin names in product_defaults
  * Fix typo in overlay imagery setup

# Output
Note: data_fusion procflow updated to use only ABI and AHI.  New output diff against original output looks correct (missing MSG coverage, and slightly different output in the overlap region (since nothing to blend with).
image

test_full_install.sh!
Package: geoips_full
Total run time: 3678 seconds
Number data types run: 50
Number data types failed: 0

Added template_basic_plugin test to test_full_install.sh after running above.  Successful direct call:
Package: my_package
Total run time: 236 seconds
Number data types run: 5
Number data types failed: 0

### CODE REVIEW

File: geoips/plugins/modules/output_formatters/metadata_tc.py [hunk 1]
```diff
@@ -97,22 +97,20 @@ def update_sector_info_with_coverage(
...
     covgs = {}
 
-    default_covg_funcs = get_covg_from_product(
-        product_name,
-        xarray_obj.source_name,
-        output_dict=output_dict,
-        covg_func_field_name="covg_func",
+    prod_plugin = products.get_plugin(xarray_obj.source_name, product_name)
+
+    default_covg_plugin = get_covg_from_product(
```
[the remark this answers is not in the document]
Comment (mindyls) [reply geoips/plugins/modules/output_formatters/metadata_tc.py:109]: I think output_dict was not used in this case (it was passed through everywhere it was actually used.  This is ok.

**Review by mindyls — APPROVED** [review]
Once all conversations resolved, I am good to go.

**Review by mindyls — APPROVED** [review]
Once all conversations resolved, I am good to go.

File: geoips/plugins/modules/output_formatters/metadata_tc.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jsolbrig) [reply geoips/plugins/modules/output_formatters/metadata_tc.py:109]: Yeah, I was careful to only drop it in places where it wasn't used. It took quite a bit of mapping.

File: geoips/schema/__init__.py [hunk 2]
```diff
@@ -1,2 +0,0 @@
-"""Schema init file."""
```
[the remark this answers is not in the document]
Comment (jsolbrig) [reply geoips/schema/__init__.py:1]: __init__.py is only needed where there is actually code. It doesn't do anything for the yaml files.

File: geoips/dev/output_config.py [hunk 3]
```diff
@@ -407,18 +406,15 @@ def get_output_formatter_kwargs(
...
-        )
-        if bg_cmap_plugin_name is not None:
-            bg_cmap_plugin = colormaps.get_plugin(bg_cmap_plugin_name)
-            bg_cmap_args = get_cmap_args(
-                bg_product_name,
-                output_formatter_kwargs["bg_xarray"].source_name,
-                output_dict=output_dict,
+        if bg_product_name:
+            bg_prod_plugin = products.get_plugin(
+                output_formatter_kwargs["bg_array"], bg_product_name
```
[the remark this answers is not in the document]
Comment (mindyls) [reply geoips/dev/output_config.py]: I fixed this I think.

File: geoips/dev/product.py [hunk 4]
```diff
@@ -822,11 +44,9 @@ def get_required_variables(product_name, source_name, output_dict=None):
             * {'<variable_type>': ['var1', 'var2', ... , 'varn']}
     """
     # This can either be a list or dictionary, dependent on YAML config specification
-    variables = get_product(product_name, source_name, output_dict=output_dict)[
```
[the remark this answers is not in the document]
Comment (jsolbrig) [reply geoips/dev/product.py:825]: THis was updated to fix this issue.

File: geoips/plugins/modules/output_formatters/imagery_annotated.py [hunk 5]
```diff
@@ -154,26 +156,22 @@ def call(
         feature_annotator=feature_annotator,
         gridline_annotator=gridline_annotator,
     )
-    product_params = None
+    prod_plugin = None
     try:
-        product_params = get_product(product_name, xarray_obj.source_name, output_dict)
-    except (KeyError, ValueError):
+        prod_plugin = products.get_plugin(xarray_obj.source_name, product_name)
```
[the remark this answers is not in the document]
Comment (jsolbrig) [reply geoips/plugins/modules/output_formatters/imagery_annotated.py]: Mindy handled this. It is working correctly now.

File: geoips/interfaces/base.py [hunk 6]
```diff
@@ -10,33 +10,210 @@
...
+    # Loop over the available schema and build a dictionary whose keys are the $ids
+    # of the schema and whose values are their associated validators. These validators
+    # can be used to validate yaml against their schema.
+    validators = {}
+    for name, schema in schema_dict.items():
+        validators[name] = validator_class(schema, registry=registry)
+    return validators
+
+
+class YamlPluginValidator:
```
[the remark this answers is not in the document]
Comment (jsolbrig) [reply geoips/interfaces/base.py:148]: Yep, it was really annoying, but moving it here solve the problem. Maybe we can find a better way to structure this later, though. I've thought about keeping the validators separate, but having any required special logic move into the plugin interfaces.

**Pull request merged**